File: algorithm/leetcode609.find-duplicate-file-in-system.cpp
class Solution {
public:
    vector<vector<string>> findDuplicate(vector<string>& paths) {
    	unordered_map<string, unordered_set<string>> m;
    	//文件内容, 文件路径集合
    	string content, path, file;
    	for(auto& p : paths) 
    	{
    		content = path = file = "";
    		int i = p.find(' ');
    		path = p.substr(0,i)+"/";
    		bool foundcontent = false;
    		for(i++; i < p.size(); ++i)
    		{
    			if(p[i] == '(')
    			{
    				foundcontent = true;
    				continue;
    			}
    			if(p[i] == ')')
    			{
    				m[content].insert(path+file);
    				foundcontent = false;
    				content = file = "";
    				i++;
    				continue;
    			}
    			if(!foundcontent)
    				file += p[i];
    			else
    				content += p[i];
    		}
    	}
    	vector<vector<string>> ans;
    	for(auto& mi : m)
    	{
    		if(mi.second.size() >= 2)
    			ans.push_back(vector<string>(mi.second.begin(), mi.second.end()));
    	}
    	return ans;
    }
};

File: algorithm/Python3_version/leetcode609.find-duplicate-file-in-system.py
class Solution:
    def findDuplicate(self, paths: List[str]) -> List[List[str]]:
        m = {};
        for p in paths:
            i = p.find(' ')
            content, path, file = "","",""
            path = p[0:i]+'/'
            foundcontent = False
            i += 1
            while i < len(p):
                if p[i]=='(':
                    foundcontent = True
                    i += 1
                    continue
                if p[i]==')':
                    if content not in m:
                        m[content] = set()
                    m[content].add(path+file)
                    foundcontent = False
                    content, file = "", ""
                    i += 2
                    continue
                if not foundcontent:
                    file += p[i]
                else:
                    content += p[i]
                i += 1
        ans = []
        for content in m:
            if len(m[content]) >= 2:
                ans.append(list(m[content]))
        return ans